dragonfly/building.py

```diff
@@ -56,6 +61,116 @@ def __init__(self, name, unique_stories):
+        stories[-1].is_top_floor = True
+
+        return cls(name, stories)
+
+    @classmethod
+    def from_all_story_geometry(cls, name, all_story_geometry, floor_to_floor_heights,
+                                tolerance):
+        """Initialize an Building from an array of Face3Ds arrays representing all floors.
+
+        This method will test to see which of the stories are geometrically unique
+        (accouting for both the floor plate geometry and the floor_to_floor_heights).
+        It will only include the unique floor geometries in the resulting Building.
+
+        All of the resulting Room2Ds will have a floor-to-ceiling height equal to the
+        Story floor-to-floor height.
+
+        Args:
+            name: Building name. Must be < 100 characters.
+            all_story_geometry: An array of arrays with each sub-array possessing
+                horizontal ladybug-geometry Face3Ds that representing the floor
+                plates of the building. Together, these Face3Ds should represent
+                all Stories of a building and each array of Face3Ds should together
+                represent one Story.
+            floor_to_floor_heights: An array of float values with a length equal
+                to the number of stories in the Building. Each value in the list
+                represents the floor_to_floor height of the Story starting from
+                the first floor and then moving to the top floor.
+            tolerance: The maximum difference between x, y, and z values at which
+                point vertices are considered to be the same. This is required as
+                a means to determine which floor geometries are equivalent to one
+                another.
+        """
+        # generate the first story of the building
+        room_2ds = cls._generate_room_2ds(
+            all_story_geometry[0], floor_to_floor_heights[0], name, 1, tolerance)
+        stories = [Story('{}_Floor1'.format(name), room_2ds, floor_to_floor_heights[0])]
+
+        # generate the remaining unique stories from the floor_to_floor_heights
+        # TODO: Add an input for core_perimeter_offsets once we have straight skeletons
+        remaining_geo = all_story_geometry[1:]
+        remaining_flr_hgts = floor_to_floor_heights[1:]
+        prev_geo = all_story_geometry[0]
+        prev_flr_to_flr = floor_to_floor_heights[0]
+        for i, (room_geo, flr_hgt) in enumerate(zip(remaining_geo, remaining_flr_hgts)):
+            # test is anything is geometrically different
+            if flr_hgt != prev_flr_to_flr or len(room_geo) != len(prev_geo) or \
+                    not all(cls._is_story_equivalent(rm1, rm2, tolerance)
+                            for rm1, rm2 in zip(room_geo, prev_geo)):
+                room_2ds = cls._generate_room_2ds(
+                    room_geo, flr_hgt, name, i + 2, tolerance)
+                stories.append(Story('{}_Floor{}'.format(name, i + 2), room_2ds, flr_hgt))
```

Comment:
This particular separator needs to remain a `_` since there's no place where I have to re-split the name again. I only use the `..` for a case where I need to eventually split the name and, for Dragonfly, that is only to denote eventual honeybee Face names generated for a Room2D.